util/types/json/serdes.go

```diff
@@ -0,0 +1,418 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package json
+
+import (
+	"bytes"
+	"encoding/binary"
+	"reflect"
+	"sort"
+	"unsafe"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/util/hack"
+)
+
```

Comment:
After compare bson and mysql binary representation, I think the last one is good enough. But if we need, it's easy for adding a prefix flag to indicate which serialize format we use.